common/trieForDirPath.go

```diff
@@ -0,0 +1,72 @@
+package common
+
+import (
+	"strings"
+)
+
+type TrieNode struct {
+	dirPathSeg             map[string]*TrieNode
+	status                 uint32
+	unregisteredButCreated bool
+	isEnd                  bool
+}
+
+type Trie struct {
```

Comment:
This is fine, but could be improved. Golang supports generics now, and if we implement this in a generic fashion, we can re-use this work to improve the scale of Sync. It's fine if this merges as is now, it's just tech debt.